### ISSUE TITLE
Does the cache get invalidated across the multi threads?

### ISSUE THREAD

**Issue by letrungkien** [svenfuchs/i18n-active_record#148]
I like the idea of this gem and would like to know if the translation cache is invalidated across the multi threads?

For example: Translation for key `hello` is `world`
- Web worker A and B are holding the cache as `{ en: { hello: 'world' } }`
- A translation update request is sent to Web worker A and updated translation for `hello` to be `zworld`, `I18n.backend.reload!` is called and Worker A will refetch the translations and now has the cache as `{ en: { hello: 'zworld' } }`

Does calling `I18n.backend.reload!` in Worker A also invalidate caches in Worker B?

**Comment by timfjord** [comment]
I don't think that the caching functionality was implemented with the multi-thread support in mind as it keeps the cache in a class variable under the hood.

**Comment by letrungkien** [comment]
Thanks. I guess it's a topic for another day.

**Comment by jaspertandy** [comment]
I believe I'm coming up against this problem now - we have an application using this, with 5 Unicorn workers. When updating a translation and calling `I18n.backend.reload!` it's only reloading translations for that worker. The current workaround that I have in place is to send a SIGHUP to the Unicorn master process to get it to reload all the workers, but I don't love this as a solution. Is there something better for rebuilding the cache across multiple processes when a change is made?

**Comment by timfjord** [comment]
As I mentioned earlier, unfortunately, the current implementation is not thread-safe, as we currently just keep the cache in a module attribute.

Alternatively, we could rely on `Thread.current` and keep the cache there.
Theoretically, it should fix the issue.

**Comment by jaspertandy** [comment]
Yeah I read what you said but I feel like almost anything would be preferable to the current implementation in a production environment. Surely the whole point of using ActiveRecord for something like this is to give users the ability to edit translations themselves without deploys needing to take place. Not giving a way to reliably purge the cache across any/all workers/sidekiq/whatever results in some bugs that are very frustrating for users.

My plan is to just implement a Redis cache that gets read/purged for our projects, but that's almost certainly insufficient for this project in the broad sense. Certainly only being able to guarantee that one worker has the latest translations at any given moment isn't enough for any production environment I run!
